pallets/circuit-breaker/src/lib.rs

```diff
@@ -0,0 +1,470 @@
+where
+	T::Balance: PartialOrd,
+{
+	pub fn update_amount(&mut self, liquidity_in: T::Balance) -> DispatchResult {
+		self.added_liquidity = self
+			.added_liquidity
+			.checked_add(&liquidity_in)
+			.ok_or(ArithmeticError::Overflow)?;
+		Ok(())
+	}
+
+	pub fn check_limit(&self) -> DispatchResult {
+		ensure!(
+			self.added_liquidity <= self.limit,
+			Error::<T>::MaxLiquidityLimitPerBlockReached
+		);
+		Ok(())
+	}
+}
+
+// Re-export pallet items so that they can be accessed from the crate namespace.
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use codec::HasCompact;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+		fn on_finalize(_n: T::BlockNumber) {
+			let _ = <AllowedTradeVolumeLimitPerAsset<T>>::clear(u32::MAX, None);
+			let _ = <AllowedLiquidityAmountPerAsset<T>>::clear(u32::MAX, None);
+		}
+
+		fn integrity_test() {
+			assert!(
+				Self::validate_limit(T::DefaultMaxNetTradeVolumeLimitPerBlock::get()).is_ok(),
+				"Circuit Breaker: Max net trade volume limit per block is set to invalid value."
+			);
+
+			if let Some(liquidity_limit) = T::DefaultMaxLiquidityLimitPerBlock::get() {
+				assert!(
+					Self::validate_limit(liquidity_limit).is_ok(),
+					"Circuit Breaker: Max liquidity limit per block is set to invalid value."
+				);
+			}
+		}
+	}
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Identifier for the class of asset.
+		type AssetId: Member
+			+ Parameter
+			+ Default
+			+ Copy
+			+ HasCompact
+			+ MaybeSerializeDeserialize
+			+ MaxEncodedLen
+			+ TypeInfo;
+
+		/// Balance type.
+		type Balance: Parameter
+			+ Member
+			+ Copy
+			+ PartialOrd
+			+ MaybeSerializeDeserialize
+			+ Default
+			+ CheckedAdd
+			+ CheckedSub
+			+ AtLeast32BitUnsigned;
+
+		/// Origin able to change the trade volume limit of an asset.
+		type TechnicalOrigin: EnsureOrigin<Self::Origin>;
+
+		/// The maximum percentage of a pool's liquidity that can be traded in a block.
+		/// Represented as a non-zero fraction (nominator, denominator) with the max value being 10_000.
+		type DefaultMaxNetTradeVolumeLimitPerBlock: Get<(u32, u32)>;
+
+		/// The maximum percentage of a pool's liquidity that can be added in a block.
+		/// Represented as an optional non-zero fraction (nominator, denominator) with the max value being 10_000.
+		/// If set to None, the limits are not enforced.
+		type DefaultMaxLiquidityLimitPerBlock: Get<Option<(u32, u32)>>;
+
+		/// Omnipool's hub asset id. The limits are not tracked for this asses.
+		type OmnipoolHubAsset: Get<Self::AssetId>;
+
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
```

Comment:
remove without_storage_info

Reply:
any idea how to bound the storage?

Reply:
it should be bounded automatically? no vecs or other dynamic structures